Solution to Python/Django Developer Test

pawel-tarasiuk opened the pull request.

pawel-tarasiuk committed:
Fixed comment formatting in url.py.

The change is small, yet makes the initial comment way more compatible with PEP 257.

pawel-tarasiuk committed:
Enabled rest_framweork in the project

pawel-tarasiuk committed:
Created analytics.models.Event model

pawel-tarasiuk committed:
Add Event model to the Django Admin panel

Event model in admin panel was also configured to be searchable by name

pawel-tarasiuk committed:
Add REST endpoint to create events, basic cleanup

Some URL configuration was moved to analytics app, homepage draft was created, /api path are set to be handled by rest_framework, /api/events was set up with entity creation via POST.

Note: created_at is listed in events serializer, as it affects listing them only. auto_now_add=True makes it impossible to post custom value anyways.

Tests yet need to be introduced.

pawel-tarasiuk committed:
Added tests for events

That includes model, API view and the API itself

pawel-tarasiuk committed:
Initial authtoken setup

Added token authentication. Custom keyword in HTTP headers, token key length (100) and foreign key name in User were configured by subclassing TokenAuthentication and Token from the REST framework.

Subclassing User would require recreation of all the migrations, so it was avoided intentionally.

pawel-tarasiuk committed:
Cleanup of authentication.py

pawel-tarasiuk committed:
Added created_by field, admin panel cleanup

This includes created_by display and searchability for events, but also some cleanup related to tokens.

pawel-tarasiuk committed:
Moved authentication to middleware

This ``middleware'' is still specific for events view. To use it with further view, the decorator_from_middleware line needs to be applied in views.py.

pawel-tarasiuk committed:
Fixed tests, cleanup

Now the tests respect the created_by field and expect the /api/events call results to depend on token.

pawel-tarasiuk committed:
Restore EventAPISerializerTests successful case

pawel-tarasiuk committed:
Added simple token tests

ilu2112 commented on analytics/views.py:
```diff
@@ -1,3 +1,30 @@
+
+@api_view(['GET', 'POST'])
+@decorator_from_middleware(middleware.AnalyticsTokenAuthMiddleware)
+def event_list(request):
```
A może można użyć widoków generycznych?

https://www.django-rest-framework.org/api-guide/generic-views/

pawel-tarasiuk replied:
Można, zrobiłem inaczej, bo było zwięźlej. Skoro jest sugerowane, to 0d50b6b4f93ab0434a462de27202d0ae66f81e6d.

Jeziersky commented on analytics/views.py:
```diff
@@ -1,3 +1,30 @@
+
+
+@api_view(['GET', 'POST'])
+@decorator_from_middleware(middleware.AnalyticsTokenAuthMiddleware)
```
https://docs.djangoproject.com/en/4.1/topics/http/middleware/#activating-middleware jeśli middleware ma być wszędzie to dlaczego nie jest wpisany do settingsów?

pawel-tarasiuk replied:
Zrobiłem to celowo, w sumie treść wskazywała tylko na /api/events. Ale to było nieco niejasne - i skoro sugestia jest taka, to 3a189efe38fe9aa81b8544c7c2f64c64be86ff08. Po to w ogóle napisałem to jako middleware, żeby dało się łatwo zmieniać jego sposób użycia.

Jeziersky commented on analytics/models.py:
```diff
@@ -1,5 +1,26 @@
-    pass
+    name = models.CharField(max_length=255)
+    created_at = models.DateTimeField(auto_now_add=True)
+    additional_data = models.TextField(blank=True, default='')
```
w takim wypadku faktycznie mamy opcjonalny tekst jednak w bazie pole nie jest nullowalne, zmieniłbym na => `additional_data = models.TextField(blank=True, null=True)`

lexpank replied:
@Jeziersky nooo, tu się nie zgodzę, `default=''` też pasi

pawel-tarasiuk replied:
@Jeziersky faktycznie, odniosłem "optional text" do procesu tworzenia eventu, ale teraz widzę, że mogło chodzić o opcjonalność w sensie języka typów. Technicznie pusty napis jest prostszy, ale takie rozumienie jest elegantsze - skoro zmiana jest sugerowana, to jasne. cb1546ce393638d729bc5df9b811586abb7c55a7 wprowadza zmianę. f624fd4914842d0ffe8329d7ec24d3a99e5df91e poprawia test, żeby pasował do zmienionego założenia.

Jeziersky commented on analytics/views.py:
```diff
@@ -1,3 +1,30 @@
+        if serializer.is_valid():
+            serializer.save(created_by=request.user)
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+    return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
zmieniłbym na HTTP_405 method not allowed ERROR

pawel-tarasiuk replied:
Przejście na generic view załatwiło sprawę, rozszerzyłem test o sprawdzenie tego dla DELETE w f624fd4914842d0ffe8329d7ec24d3a99e5df91e

Jeziersky commented on analytics/views.py:
```diff
@@ -1,3 +1,30 @@
+        serializer = serializers.EventSerializer(events, many=True)
+        return Response(serializer.data)
+    if request.method == 'POST':
+        serializer = serializers.EventSerializer(data=request.data)
```
- serializer przy tworzeniu przyjmuje created_at (sql i tak wstawi swoją date, ale można było to rozdzielić na EventGetSerializer oraz EventCreateSerializer)

pawel-tarasiuk replied:
Można, celowo nie rozdzielałem. Korzyści z takich decyzji byłyby potencjalne, nie dotyczyłoby wprost realizacji punktów wprost zapisanych w zadaniu. Działanie jest poprawne, a domyślne rozwiązania są moim zdaniem czytelniejsze (szczególnie przy generic view, gdzie get_serializer w tym przypadku zawsze wystarczy).

Jeziersky commented on analytics/admin.py:
```diff
@@ -1,3 +1,37 @@
+# Add the adequate auth tokens.
+@admin.register(authentication.AnalyticsToken)
+class AnalyticsTokenAdmin(admin.ModelAdmin):
+    list_display = ('user', 'key', )
```
możaby dociągać usera odrazu w get_queryset (oszczędzi to zapytań do bazy)

pawel-tarasiuk replied:
Proponuję dociągnięcie jak w db514b459e4cc1bb34c537ee8cb0d8179a1cbc48, orm powinien wtedy zapewnić oszczędność w liczbie zapytań

Jeziersky commented on analytics/authentication.py:
```diff
@@ -0,0 +1,41 @@
+    @classmethod
+    def generate_key(cls):
+        """Generate 100 characters long key."""
+        return binascii.hexlify(os.urandom(50)).decode()
```
co jeśli wartość powtórzy się w bazie? user zostanie stworzony bez tokenu? user nie zostanie stworzony?

pawel-tarasiuk replied:
Poleci IntegrityError. Poprawiłem w 19e9aa59cee0f5c2a6c1683c33f40a02bb82b428.

pawel-tarasiuk committed:
Reimplemented events view using generics

pawel-tarasiuk committed:
Enabled middleware globally

Instead of per-view decorators, not it is enabled in settings.py. NOTE: instead of specific authentication for /api/events now it is classical middleware.

pawel-tarasiuk committed:
Minor tests cleanup

pawel-tarasiuk committed:
Event additional_data is nullable

pawel-tarasiuk committed:
Fixed tests, added unallowed method testing

pawel-tarasiuk committed:
Joined user to token queryset in admin panel

pawel-tarasiuk committed:
Catch possible IntegrityError on duplicate tokens

Due to the birthday problem, probability of this happening at least once would reach 50% at about 3.5e60 users.

pawel-tarasiuk committed:
Cleanup
